### ISSUE TITLE
ReactorTime has serious performance issues

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10455]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| stefan reported|
|-|-|
|Trac ID|trac#707|
|Type|defect|
|Created|2004-09-09 15:33:08Z|

Attachments:

* [base.py.diff.gz](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/c9cda041d19268b1e4af5ec9807de1c5a06b8b0a.gz) (2140 bytes) - added by stefan on 2004-09-09 15:33:07Z - 
* [base.py.diff.2.gz](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/386a2f8e10875bb98ece6da3509caa9ca0798c2e.gz) (2235 bytes) - added by stefan on 2004-09-09 15:52:21Z - 
* [base.py.diff.3.gz](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/ac80de4a802e6dd94b3bc323b2c208eb573e2af7.gz) (2564 bytes) - added by stefan on 2004-09-09 18:12:02Z - 
* [base.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/3ea08dbcdcaecc22c9eed96f9e27c8c2f46a7711.diff) (7136 bytes) - added by jknight on 2004-09-10 02:00:38Z - 
* [ReactorTimeTest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/03c063435b909f4432ec9381bd4105db92327670.py) (2154 bytes) - added by stefan on 2004-10-07 21:20:12Z - 
* [ReactorTimeBenchmark.060sec.LIST2.stats](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/1f22b74ed91608360aec0efc5db49ec2bdf8e4fc.stats) (12174 bytes) - added by stefan on 2004-10-07 21:22:16Z - 
* [ReactorTimeBenchmark.060sec.HEAP2.stats](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/de594c84b11690733a9c329c2cc3e1bc5f70e97d.stats) (12952 bytes) - added by stefan on 2004-10-07 21:22:54Z - 
* [base.py.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2a8/2a8ae2469569e6eccde1a5b5b16eb076d0769ad3/31a06936b498faafc0512862d6f049288fb5c791.diff) (7014 bytes) - added by stefan on 2004-10-07 21:28:28Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__707 707
type__defect defect
reporter__stefan stefan
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1094743988000000 1094743988000000
changetime__1099953910000000 1099953910000000
version__ 
owner__ 
cc__exarkun cc__itamarst cc__jknight cc__stefan
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
As discussed on the mailing list, here is a patch for base.py that replaces the
current list based IReactorTime implementation with a heap based one. It also
moves a lot of the work into runUntilCurrent in order to speed up calls to
callLater, cancel, reset and delay.

The complexity of cancel, reset and delay moves from linear time in the original
implementation to constant time in many cases and logarithmic time in others.

Some obviously positive improvements from the mailing list discussion are
already integrated, including constant time for delay. cancel() takes
logarithmic time which is a trade off with the runtime of runUntilCurrent.
cancel() is only called once for a delayed call, but the performance of
runUntilCurrent suffers from each cancelled call that stays on the heap.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Another patch that improves the performance of reset in the case where the new
delay is longer than the old one. In that case reset is the same as delay, i.e.
constant time.

I appended the new (complete) patch for convenience.


--- base.py.new 2004-09-09 11:41:31.329751790 +0200
+++ base.py     2004-09-09 11:43:59.037635320 +0200
@@ -99,8 +99,13 @@
         elif self.called:
             raise error.AlreadyCalled
         else:
-            self.time = seconds() + secondsFromNow
-            self.resetter(self)
+            new_time = seconds() + secondsFromNow
+            if new_time &lt; self.time:
+                self.delayed_time = 0
+                self.time = new_time
+                self.resetter(self)
+            else:
+                self.delayed_time = new_time - self.time

     def delay(self, secondsLater):
         """Reschedule this call for a later time
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Another complete patch with a rewrite of reactor.timeout()

Timeout had to be adapted to delayed calls on the heap. It wasn't really broken
in the last version, but the timeout returned by this method could be lower than
necessary. So the current version now resolves as many delayed calls as we find
at the top of the heap and only then does it calculate the timeout.

DelayedCall.getTime() was broken after calling delay or reset.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Modification of the stefan's patch:

Make cancel constant time, but don't allow left-over cancelled items to become a huge majority 
of the heap.
  - On cancel(), set a flag, and increment a counter of cancelled items in the list.
  - in runUntilCurrent, if the number of cancelled items in the list is both > 50 and > 1/2 of the total 
items, filter them out and re-heapify the list.

Remove custom implementation of heapq
  - This means moving a DelayedCall to happen *sooner* that it was previously scheduled for now 
causes a list.find operation to happen, which of course takes linear time in the list length. I posit that 
moving a DelayedCall to happen sooner almost never happens so this doesn't matter.
  - Now we can use builtin heapq which is in C in Py2.4
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Ok, I think that's about Ok. I guess you forgot to test the new re-heapify part,
though. It misses some "self."s and an underscore. The last part of the patch
has to look like this:

--------------------------------
@@ -428,7 +491,14 @@
                     e += "".join(call.creator).rstrip().replace("\n","\n C:")
                     e += "\n"
                     log.msg(e)
-
+
+        if (self._cancellations > 50 and
+             self._cancellations > len(self._pendingTimedCalls) // 2):
+            self._cancellations = 0
+            self._pendingTimedCalls = [x for x in self._pendingTimedCalls
+                                       if not x.cancelled]
+            heapify(self._pendingTimedCalls)
+


     # IReactorThreads
--------------------------------

Consider running pychecker.

Then, I'd actually like to see this patch integrated as soon as possible. I
consider this a huge speed up for this part of the reactor.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Stefan: This patch will almost certainly be accepted if it is accompanied with some benchmark 
code comparing the speed of operations in the new code to that of the current code. Tests 
showing both what got faster and what got slower would be extra helpful.

I would do this eventually, but I think you already have some anyways?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Ok, I did a bit of testing with the little program I attached. One of
the problems I found is that the original reactor does not terminate
in some cases. That's because of the infinite-loop problem that Bob
Ippolito mentioned on the mailing list. So we may even consider the
current implementation broken as it can lead to non-responsiveness to
I/O and other events. I had to change the test to use 1 second offset
minimum instead of 0. This problem is solved in the current patch, but
that problem is actually independant of the intention of this
patch. Anyway...

What it does is: run 100-10000 delayed calls in parallel and in each
of them cancel, reset or delay one of the others. Cancel is called
more often than delay which is called more often than reset, I
considered that a relevant application case. That would even make a
good stress case, I guess...

It does this for 60 seconds, collects the stats from the profiler and
increases the number of parallel calls. I ran it on Python-2.4a3 to
see how it will perform in the future. From the numbers I see I expect
the gain in 2.3 to be much bigger.

What I found is that runUntilCurrent is called much more often with
the new version, which is due to the delayed merge of new calls (as
Bob suggested). The old version simply added them on request, the new
version returns fron runUntilCurrent first and merges them during the
next run. In my specific example, where each running call adds a new
one, this makes a huge difference and adds the overhead of running
through the rest of the reactor before starting the next call. This
makes pure ReactorTime programs slower and adds responsiveness to I/O
apps. As I said, the old way of doing it was a bug, so this should be
the right trade-off. However, it makes the execution times of
runUntilCurrent incomparable between the two versions.

Another thing I could see from the traces is that Python-2.4 totally
rocks. The results I see here with the original version are so much
faster than my early tests with 2.3! That was what made me discontinue
any use of the cancel() funktion. I didn't run the tests with 2.3 now,
but I'd expect the gain to be even more substantial there.

For the comparison I may add that the original version reached 100%
CPU already at the 3000 calls step while the new implementation did
that from 6000 on. I didn't compare in between, so maybe its only
3000/5000, but the comparison at 3000 looked good enough.

What you can see from the traces is that the gain for 100 concurrent
calls is lower than what runUntilCurrent eats up now. So you should
expect to loose some milliseconds here. At 1000 the gain is already
visible (mind the higher number of function calls) and it then
increases to a factor of more than 2 at 10000 concurrent calls.

You can also see the improvement for callLater, cancel, reset and
delay. All of them got considerably faster, by factors.

I'm actually sorry that the numbers are not directly comparable. You
could achieve that by calling reactor.stop from within the chooser
function after a fixed number of runs, but that number then depends on
the call rate, and ... well ...

I hope evidence is satisfied anyway.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Profiler results with the old list based implementation
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Profiler results with the new heap based implementation
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
Current patch, contains fixes for typos I mentioned last time. Also: list.find()
doesn't exist, so I had to replace it with list.index and catch a possible
exception. I do not actually know why this exception can be raised, so maybe
there's a little bug hidden somewhere, but in any case it doesn't harm to just
catch it and go on.

Didn't Python's Zen mention that exceptions shouldn't pass silent - unless
explicitely silenced? Well, this one is silence enough...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Exarkun: any objections?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Committed, r12283.
Made a couple changes: timeout() inserts new timed calls, and then only looks at the first item: doesn't 
go through looking for cancellations. The runUntilCurrent loop was looking at .getTime() when it should 
have been looking at .time.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
One note on the silenced exception raised by heap.find() that I mentioned in
msg2697: I'm now convinced that it occurs when the delayed call that is being
canceled has not yet been merged into the heap, i.e. is still pending in the
list of new calls.

Therefore, as I figured before and as it is implemented in the current version,
it is perfectly correct to just silence it since the canceled call will be
removed during the next iteration anyway.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="stefan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefan commented|
|-|-|

```
#!html
<pre>
changed status back to resolved (seems to have been reset to chatty automatically)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
